core/templates/services/request-interceptor.service.ts

```diff
@@ -80,11 +80,11 @@ export class RequestInterceptor implements HttpInterceptor {
         .pipe(
           switchMap(token => {
             if (request.method === 'POST' || request.method === 'PUT') {
-              var body = new HttpParams()
-                // @ts-ignore
-                .set('csrf_token', token)
-                .set('source', document.URL)
-                .set('payload', JSON.stringify(request.body));
+              var body = new FormData();
```

Comment:
Just to check -- you sure about this? It touches all HTTP requests, not just the image one.

Reply:
This was not meant as a fix for images. There seems to be an issue when you try to add a reserved character in the body of any POST call routed through this service. ie. Adding ";" to the topic name or any other field in the new structures will throw a 500 error. This change should fix that.

Reply:
Sounds good, thanks. Could you please make sure there are unit tests for this? (Always do that as a best practice, otherwise things may regress in the future.)

Reply:
Hi @seanlip, I tried adding a spec for testing the request-interceptor.service because it didn't have one yet but I was unsuccessful. I ran into issues while testing the interceptor for post and put requests (which is required to test my code changes). It works fine for GET requests. The expectOne method fails to match the URL I specify for the POST request. 

<details><summary>Code</summary><p>

```ts
import { TestBed,  fakeAsync, flushMicrotasks } from '@angular/core/testing';
import {
  HttpClientTestingModule,
  HttpTestingController
} from '@angular/common/http/testing';
import {HTTP_INTERCEPTORS, HttpClient} from '@angular/common/http';

import {
  MockCsrfTokenService,
  RequestInterceptor
} from 'services/request-interceptor.service';

describe('Request Interceptor Service', () => {
  let ris: RequestInterceptor = null;
  let mcts: MockCsrfTokenService = null;
  let httpClient: HttpClient = null;
  let httpTestingController: HttpTestingController = null;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [HttpClientTestingModule],
      providers: [{
        provide: HTTP_INTERCEPTORS,
        useClass: RequestInterceptor,
        multi: true
      }]
    });
    ris = TestBed.get(RequestInterceptor);
    mcts = TestBed.get(MockCsrfTokenService);
    httpClient = TestBed.get(HttpClient);
    httpTestingController = TestBed.get(HttpTestingController);
    spyOn(mcts, 'getTokenAsync').and.callFake(() => {
      return new Promise((resolve) => {
        resolve('sample-csrf-token');
      });
    });
  });

  afterEach(() => {
    httpTestingController.verify();
  });

  it('should expect request body to be a FormData constructor',
    fakeAsync(() => {
    httpClient.post('/api', {data: 'test'}).subscribe(
      response => expect(response).toBeTruthy());
    let req = httpTestingController.expectOne('/api');
    expect(req.request.method).toEqual('POST');
    req.flush({data: 'test'});
    flushMicrotasks();
  }));
});
```
Error:
`Error: Expected one matching request for criteria "Match URL: /api", found none.`

</p>
</details>

Here are the resources I used to write this code ([ref1](https://alligator.io/angular/testing-http-interceptors/), [ref2](https://stackoverflow.com/a/46698139)). On realising that we use expectOne for testing several services that use POST calls, I figured it might be because I'm not using a service to trigger a POST call and decided to add a test to an existing service to check if the interceptor works correctly. I tried following the tutorial [here](https://alligator.io/angular/testing-http-interceptors/) and added some code to skill-creation-backend-api.service so that POST calls pass through the interceptor. However, this seemed to fail too with a similar error:

`Error: Expected one matching request for criteria "Match URL: /skill_editor_handler/create_new", found none.`

<details><summary>Code</summary><p>

![image](https://user-images.githubusercontent.com/11008603/81055954-5b629b80-8ee7-11ea-9209-294a695f3156.png)

</p></details>

I'm stuck on this currently. I could write an e2e test to cover this scenario and ensure that this  doesn't regress. Thoughts?

Reply:
Hi @marianazangrossi, can you take a quick look and see what I might be doing wrong?

Reply:
Tagging @bansalnitish as well, in case he knows what may be going wrong with the interceptor.

Reply:
Hi @kevintab95! After searching a bit, I've found that what you're facing is an issue from Angular. The problem is this piece of code from `request-interceptor.service.ts` (on line 79):

```
return from(this.csrf.getTokenAsync())
        .pipe(
          switchMap(token => {
```

Based on this [issue](https://github.com/angular/angular/issues/25733) and this [repository](https://github.com/seangwright/ng-promise-interceptor-test-bug), when passing a promise as parameter in `from` method and using a rxjs pipeline, it will make the unit tests unstable.

So here's what I did:
- Instead of returning a promise in `getTokenAsync` mock, I replaced it by an array (rxjs from method only works with arrays, promises and iterables -- as you can see [here](https://www.learnrxjs.io/learn-rxjs/operators/creation/from)).
- Removed the `fakeAsync` and `flushMicrotasks` methods.

Here's is the code I did (and it passed the tests):
```
import { TestBed } from '@angular/core/testing';
import {
  HttpClientTestingModule,
  HttpTestingController
} from '@angular/common/http/testing';
import {HTTP_INTERCEPTORS, HttpClient} from '@angular/common/http';

import {
  MockCsrfTokenService,
  RequestInterceptor
} from 'services/request-interceptor.service';

fdescribe('Request Interceptor Service', () => {
  let mcts: MockCsrfTokenService = null;
  let httpClient: HttpClient = null;
  let httpTestingController: HttpTestingController = null;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [HttpClientTestingModule],
      providers: [{
        provide: HTTP_INTERCEPTORS,
        useClass: RequestInterceptor,
        multi: true
      }]
    });

    mcts = TestBed.get(MockCsrfTokenService);
    httpClient = TestBed.get(HttpClient);
    httpTestingController = TestBed.get(HttpTestingController);
    spyOn(mcts, 'getTokenAsync').and.returnValue(['sample-csrf-token']);
  });

  afterEach(() => {
    httpTestingController.verify();
  });

  it('should expect request body to be a FormData constructor', () => {
    httpClient.post('/api', {data: 'test'}).subscribe(
      response => expect(response).toBeTruthy());

    let req = httpTestingController.expectOne('/api');
    expect(req.request.method).toEqual('POST');
    req.flush({data: 'test'});
  });
});
```

Reply:
Thank you so much @marianazangrossi!! I'll try this out! :)

Reply:
You're welcome!